Library/Homebrew/cmd/upgrade.rb

```diff
@@ -171,4 +174,93 @@ def upgrade_formula(f)
       nil
     end
   end
+
+  def relink_dependents(formula)
+    broken_formulae = []
+
+    CacheStoreDatabase.use(:linkage) do |db|
+      kegs = Formula.installed
+                    .collect(&:opt_or_installed_prefix_keg)
+                    .reject { |f| f.runtime_dependencies.nil? }
+                    .keep_if { |f| f.runtime_dependencies.include?(formula) }
```

Comment:
thanks; it's getting closer. the following makes it work for me in my testing:

~~~
diff --git a/Library/Homebrew/cmd/upgrade.rb b/Library/Homebrew/cmd/upgrade.rb
index 94a99c0..1f86258 100644
--- a/Library/Homebrew/cmd/upgrade.rb
+++ b/Library/Homebrew/cmd/upgrade.rb
@@ -182,7 +182,7 @@ module Homebrew
       kegs = Formula.installed
                     .collect(&:opt_or_installed_prefix_keg)
                     .reject { |f| f.runtime_dependencies.nil? }
-                    .keep_if { |f| f.runtime_dependencies.include?(formula) }
+                    .keep_if { |f| f.runtime_dependencies.collect{ |d| d["full_name"] == formula.full_name }.any? }
 
       odebug "Checking dynamic library links of #{formula}'s dependents" unless kegs.empty?
 
@@ -192,7 +192,7 @@ module Homebrew
 
         odebug "#{keg} has broken dynamic library links and will be reinstalled from source"
 
-        broken_formulae << Formulary.from_keg(f)
+        broken_formulae << Formulary.from_keg(keg)
       end
     end
 
~~~